functional_tests/steps/bundles.py

```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+        context.users = {}
+    context.users[user_role] = create_user(user_role)
+
+
+@given("there is a Statistical Analysis page")
+def statistical_analysis(
+    context: Context,
+) -> None:
+    context.article_series_page = ArticleSeriesPageFactory(title="PSF")
+    article = StatisticalArticlePageFactory(parent=context.article_series_page)
+    mark_page_as_ready_to_publish(article, UserFactory())
+    context.statistical_article_page = article
+
+
+@given("there is a release calendar page")
+def release_calendar(context: Context) -> None:
+    nowish = timezone.now() + timedelta(minutes=20)
+    context.release_calendar_page = ReleaseCalendarPageFactory(
+        release_date=nowish, title="Release Calendar Page bundles UI", status=ReleaseStatus.PROVISIONAL
+    )
+
+
+@given("there is a preview team")
+def preview_teams_create(context: Context) -> None:
+    context.team = Team.objects.create(identifier="preview-team", name="Bundles UI Preview team", is_active=True)
+
+
+@given("the {user_role} is a member of the preview team")
+def add_user_to_preview_teams(context: Context, user_role: str) -> None:
+    user = context.users[user_role]["user"]
+    user.teams.add(context.team)
+
+
+@given("there are {number_of_bundles} bundles with {bundle_details}")
+def multiple_bundles_create(context: Context, number_of_bundles: str, bundle_details: str) -> None:
+    bundle_dets = json.loads(bundle_details)
+    if bundle_dets:
+        context.bundles = []
+        context.bundlepages = []
+        context.bundleteams = []
+
+        for __ in range(int(number_of_bundles)):
+            bundle = BundleFactory()
+
+            if not hasattr(context.users, bundle_dets["creator_role"]) and bundle_dets["creator_role"] != "":
+                context.users[bundle_dets["creator_role"]] = create_user(bundle_dets["creator_role"])
+
+            bundle_status = BundleStatus.DRAFT
+            bundle_approved = False
+            if bundle_dets["status"] == "Approved":
+                bundle_status = BundleStatus.APPROVED
+                bundle_approved = True
+            elif bundle_dets["status"] == "In_Review":
+                bundle_status = BundleStatus.IN_REVIEW
+
+            if bundle_dets["creator_role"] != "":
+                user = context.users[bundle_dets["creator_role"]]["user"]
+
+                bundle.created_by = user
+                bundle.status = bundle_status
+                bundle.approved = bundle_approved
+
+            else:
+                bundle.status = bundle_status
+                bundle.approved = bundle_approved
+            bundle.save()
+
+            if bool(bundle_dets["preview_teams"]) and hasattr(context, "team"):
+                BundleTeam.objects.create(parent=bundle, team=context.team)
+                context.bundleteams.append({"parent": bundle, "team": context.team})
+            bundle.save()
+
+            if bool(bundle_dets["add_rel_cal"]) and hasattr(context, "release_calendar_page"):
```

Comment:
If your bundle details have `"add_rel_cal"` as true, but you forgot to add a step to add a release calendar, this indicates a bug in the test code. Don't check for `hasattr(context, "release_calendar_page")`, fail here so that you can see your mistake.